### PR TITLE
[FLINK-33192] Fix Memory Leak in WindowOperator due to Improper Timer Cleanup

### PR DESCRIPTION
**Pull request by kartikeypant**
## What is the purpose of the change
* This Pull Request addresses an important state memory leak issue identified within the default window operator of Apache Flink. 
* After this change, the cleanup timer should be registered for every window that's added to the window state regardless of it emitting a result after it’s fired.
* This change is associated to the following bug: [FLINK-33192
](https://issues.apache.org/jira/projects/FLINK/issues/FLINK-33192)

## Brief change log
Modified `org.apache.flink.streaming.runtime.operators.windowing.WindowOperator` to register clean up timer for window states regardless of whether emits a result or not. 

## Verifying this change
- WindowOperatorTest is working for functional correctness.
- Written a new test `testCleanupTimerWithEmptyStateNoResultForTumblingWindows`. This test simulates a empty returning aggregate function with tumbling windows. The passing of the test validates that the window contents are being cleared during the time the cleanup time is actually supposed to be called. Hence, the memory leak should not be happening after the change. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by fredia** [comment]
@kartikeypant Thanks for the PR, overall LGTM, would you like to add a test about this in `WindowOperatorTest`?

**Comment by kartikeypant** [comment]
Hey @fredia, thank you for your feedback! 

I have added the test `testCleanupTimerWithEmptyStateNoResultForTumblingWindows` (as specified) in WindowOperatorTest and updated the description of the PR with the same. 

Could you please review the changes and let me know if any further adjustments are needed?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java [hunk 1]
```diff
@@ -3130,6 +3130,113 @@ void testCleanupTimerWithEmptyReduceStateForSessionWindows() throws Exception {
...
+        expected.add(new Watermark(1699));
+        expected.add(new Watermark(1799));
+        expected.add(new Watermark(1999)); // here it fires and purges
+        expected.add(new Watermark(2000)); // here is the cleanup timer
+        expected.add(new Watermark(5000));
+
+        System.out.println(expected);
+        System.out.println(testHarness.getOutput());
+        TestHarnessUtil.assertOutputEqualsSorted(
+                "Output was not correct.", expected, expected, new Tuple2ResultSortComparator());
```
Comment (fredia) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java:3236]:
```suggestion
                "Output was not correct.", expected, testHarness.getOutput(), new Tuple2ResultSortComparator());
```

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java [hunk 2]
```diff
@@ -3130,6 +3130,113 @@ void testCleanupTimerWithEmptyReduceStateForSessionWindows() throws Exception {
...
+                .isEqualTo("[]");
+
+        expected.add(new Watermark(1599));
+        expected.add(new Watermark(1699));
+        expected.add(new Watermark(1799));
+        expected.add(new Watermark(1999)); // here it fires and purges
+        expected.add(new Watermark(2000)); // here is the cleanup timer
+        expected.add(new Watermark(5000));
+
+        System.out.println(expected);
```
Comment (fredia) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorTest.java:3233]: nit: remove ` System.out.println()`

**Review by fredia — CHANGES_REQUESTED** [review]
Thanks for the update, overall LGTM, I left two minor comments.

And it's better to rebase these commits to master, and squash commits into one.

**Review by fredia — APPROVED** [review]
Thanks for the update, let's wait the CI green.
And would you like to backport this PR to 1.18/1.19/1.20? something like https://issues.apache.org/jira/browse/FLINK-34274

**Comment by kartikeypant** [comment]
Sure, thanks @fredia for the guidance - will back-port these changes to v1.18/1.19/1.20 as well as suggested.

**Comment by fredia** [comment]
@flinkbot run azure

**Pull request merged**